general/echo/kmdf/exe/src/main.rs

```diff
@@ -0,0 +1,555 @@
+#![deny(rustdoc::missing_crate_level_docs)]
+#![deny(rustdoc::invalid_codeblock_attributes)]
+#![deny(rustdoc::invalid_html_tags)]
+#![deny(rustdoc::invalid_rust_codeblocks)]
+#![deny(rustdoc::bare_urls)]
+#![deny(rustdoc::unescaped_backticks)]
+#![deny(rustdoc::redundant_explicit_links)]
+
+use std::{env, error::Error, ffi::OsString, os::windows::prelude::*, sync::RwLock, thread};
+
+use once_cell::sync::Lazy;
+use uuid::{uuid, Uuid};
+use windows_sys::Win32::{
+    Devices::DeviceAndDriverInstallation,
+    Foundation::{CloseHandle, GetLastError, BOOL, ERROR_IO_PENDING, HANDLE, INVALID_HANDLE_VALUE},
+    Storage::FileSystem::{
+        CreateFileW,
+        ReadFile,
+        WriteFile,
+        FILE_FLAG_OVERLAPPED,
+        FILE_GENERIC_READ,
+        FILE_GENERIC_WRITE,
+        FILE_SHARE_READ,
+        FILE_SHARE_WRITE,
+        OPEN_EXISTING,
+    },
+    System::{
+        Threading::INFINITE,
+        IO::{CreateIoCompletionPort, GetQueuedCompletionStatus, OVERLAPPED, OVERLAPPED_0},
+    },
+};
+
+#[derive(Default, Debug)]
+struct Globals {
+    perform_async_io: bool,
+    limited_loops: bool,
+    async_io_loops_num: usize,
+    device_path: String,
+}
+
+static GLOBAL_DATA: Lazy<RwLock<Globals>> = Lazy::new(|| RwLock::new(Globals::default()));
+static GUID_DEVINTERFACE_ECHO: Uuid = uuid!("CDC35B6E-0BE4-4936-BF5F-5537380A7C1A");
+static READER_TYPE: u32 = 1;
+static WRITER_TYPE: u32 = 2;
+static NUM_ASYNCH_IO: usize = 100;
+static BUFFER_SIZE: usize = 40 * 1024;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let argument_vector: Vec<String> = env::args().collect();
+    let argument_count = argument_vector.len();
+
+    if argument_count > 1 {
+        if argument_vector[1] == "-Async" {
+            let mut globals = GLOBAL_DATA.write()?;
+            globals.perform_async_io = true;
+            if argument_count > 2 {
+                globals.async_io_loops_num = argument_vector[2].parse::<usize>()?;
+                globals.limited_loops = true;
+            } else {
+                globals.limited_loops = false;
+            }
+        } else {
+            eprintln!(
+                r##"
+Usage:
+    Echoapp.exe         --- Send single write and read request synchronously
+    Echoapp.exe -Async  --- Send reads and writes asynchronously without terminating
+    Echoapp.exe -Async <number> --- Send <number> reads and writes asynchronously
+Exit the app anytime by pressing Ctrl-C
+"##
+            );
+            return Err("Invalid Args".into());
+        }
+    }
+
+    get_device_path(&GUID_DEVINTERFACE_ECHO)?;
+
+    let globals = GLOBAL_DATA.read()?;
+    println!("DevicePath: {}", globals.device_path);
+    let mut path_vec = globals.device_path.encode_utf16().collect::<Vec<_>>();
+    let perform_async_io = globals.perform_async_io;
+    drop(globals);
+
+    let h_device: HANDLE;
+    path_vec.push(0);
+    let path = path_vec.as_ptr();
+
+    // SAFETY:
+    // Call Win32 APIs to access driver
```

Comment:
The point of enforcing `SAFETY` comments is that they should justify why the code in the unsafe block is **guaranteed** to be safe (since the compiler cannot statically guarantee it in `unsafe` blocks). This can then be used when auditing unsafe blocks.

In order to provide that justification, we must understand exactly why the unsafe block is needed. In this block, there are two reasons: 
1. the use of raw pointers: any function that accepts a raw pointer must be marked as an `unsafe` function since the function may dereference a pointer (which is an unsafe op). This block needs to justify why this pointer is _valid_ per [ptr safety](https://doc.rust-lang.org/std/ptr/index.html#safety)(ie. non-null, aligned, points to valid memory, is dereferenceable per the ptr safety docs, remains valid for entirety of function etc.).

2. calling into FFI functions: calling into FFI is inherently unsafe. Any calls to FFI functions must guarantee that the call (including its arguments), are guaranteed to abide by the contract set by the API. In this case, the docs for CreateFileW must read and the comment should say how all the api requirements outlined in the docs are upheld. If possible, have some way to guarantee they are upheld (ex. compile time assert).

See [this pr](https://github.com/microsoft/windows-drivers-rs/pull/24/files#diff-d0dc5143131ae33b24f02dd164505ab806c7136bc7822e77ff7d74152f7082e3) for some examples of useful safety comments

See [this rust mdbook](https://std-dev-guide.rust-lang.org/policy/safety-comments.html#:~:text=Safety%20comments%20Using%20unsafeblocks%20is%20often%20required%20in,Below%20are%20some%20examples%20taken%20from%20the%20standard) for more about safety comments.

Reply:
Updated comments